### PR TITLE
Enhance the support for BYTES data type

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
1. Add BytesUtils to handle encode/decode for Hex string
2. Add FieldSpec.DataType.convert() to convert string value to data type
3. Support Hex string for index(), indexOf(), inRange() in all BYTES dictionaries
4. Add MutableDictionary.compare(dictId1, dictId2) to help the sort process and support sorting byte[]
5. Support returning Hex string in group-by for byte[] column
6. In MutableSegmentImpl, uniform the sort for all types, reduce the memory usage especially for STRING and BYTES columns

### CODE REVIEW

File: pinot-common/src/main/java/org/apache/pinot/common/data/FieldSpec.java [hunk 1]
```diff
@@ -439,6 +415,28 @@ public int size() {
           throw new IllegalStateException("Cannot get number of bytes for: " + this);
       }
     }
+
+    /**
+     * Converts the given string value to the data type.
+     */
+    public Object convert(String value) {
```
Comment (kishoreg) [inline pinot-common/src/main/java/org/apache/pinot/common/data/FieldSpec.java]: probably not part of this PR but why is this method part of the FieldSpec?

File: pinot-common/src/main/java/org/apache/pinot/common/utils/BytesUtils.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+   */
+  public static byte[] toBytes(Object rawValue) {
+    if (rawValue instanceof String) {
+      try {
+        return Hex.decodeHex(((String) rawValue).toCharArray());
+      } catch (DecoderException e) {
+        throw new IllegalArgumentException("Value: " + rawValue + " is not Hex encoded", e);
+      }
+    } else {
+      return (byte[]) rawValue;
```
Comment (kishoreg) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/BytesUtils.java]: might need additional safety check if it can be cast to byte[]

File: pinot-common/src/main/java/org/apache/pinot/common/utils/primitive/ByteArray.java [hunk 3]
```diff
@@ -45,19 +45,13 @@ public int length() {
...
-  /**
-   * Static utility function to convert a byte[] to Hex string.
-   *
-   * @param bytes byte[] to convert
-   * @return Equivalent Hex String.
-   */
-  public static String toHexString(byte[] bytes) {
-    return Hex.encodeHexString(bytes);
+  public String toHexString() {
+    return BytesUtils.toHexString(_bytes);
```
Comment (kishoreg) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/primitive/ByteArray.java]: why did we move this to BytesUtils class?

File: pinot-core/src/main/java/org/apache/pinot/core/query/pruner/AbstractSegmentPruner.java [hunk 4]
```diff
@@ -87,26 +89,17 @@ protected boolean pruneNonLeaf(@Nonnull FilterQueryTree filterQueryTree,
...
-          return Long.valueOf(input);
-        case FLOAT:
-          return Float.valueOf(input);
-        case DOUBLE:
-          return Double.valueOf(input);
-        case STRING:
-          return input;
-        default:
-          throw new IllegalStateException("Unsupported data type: " + dataType);
+      if (dataType != FieldSpec.DataType.BYTES) {
```
Comment (kishoreg) [inline pinot-core/src/main/java/org/apache/pinot/core/query/pruner/AbstractSegmentPruner.java]: Is there a way we can eliminate this special check for BYTES?

File: pinot-common/src/main/java/org/apache/pinot/common/data/FieldSpec.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/data/FieldSpec.java]: This is part of the DataType, which is inside FieldSpec.

File: pinot-common/src/main/java/org/apache/pinot/common/utils/BytesUtils.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/BytesUtils.java]: If it cannot be cast, it will throw ClassCastException, which is the expected behavior.

File: pinot-common/src/main/java/org/apache/pinot/common/utils/primitive/ByteArray.java (hunk 3 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/primitive/ByteArray.java]: BytesUtils is the util class for all byte[] related methods. ByteArray is the comparable wrapper for byte[], which should not be used as util class.

File: pinot-core/src/main/java/org/apache/pinot/core/query/pruner/AbstractSegmentPruner.java (hunk 4 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/query/pruner/AbstractSegmentPruner.java]: If we only keep ByteArray as the value inside Pinot, we can achieve that. Right now we have mixed of byte[] and ByteArray, thus hard to uniform all behaviors. We might want to change the code to use only ByteArray internal in a separate PR.

Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/query/pruner/AbstractSegmentPruner.java]: Internally we mix the use of byte[] and ByteArray, which causes this special handling. IMO we should uniform the behavior to always use ByteArray (at least for segment generation, and filters where comparison applies). That will involve much bigger changes, which can be submitted in a separate PR if we decide to do so.
